### PR TITLE
revive STAGE2 RISCV64 backend

### PR DESCRIPTION
**Pull request by Rexicon226**
@kubkon
only took me a month of work to feel this is enough to upstream.

- fixed a ton of incorrect stuff.
- struct / array access when they are on the stack, (still need to figure out the shifting stuff for registers).
- globals loading through the `load_symbol`.
- `@abs()` for some cases.
- `@byteSwap()` for 16-bit.
- loading 64-bit immediate (although in a bad way, really should not being allocating a register to do it).
- multiple args work correctly now, even though multiple levels of functions.
- pass by reference (aka pointer args) work now, and can be assigned to and stuff.
- lots of boilerplate added for addition and comparing and shifting.
- actual overflow checks for `airAddWithOverflow` using the struct parsing stuff.
- slices work now, and we have panic messages with messages.
- added `R_ZIG_GOT_HI20` and `R_ZIG_GOT_LO12` for loading symbols from `.got.zig`.
- we actually follow the RISC-V calling convention correctly now.
- lots and lots more

all of these small little bug fixes will add up to create a nice stable environment for other contributors who are interested in RISC-V to find it more fun. not saying we're there yet, i'm rewriting stuff every hour.

PS. up to you guys if you wanna keep the panic function i made for `stage2_riscv64`. i'd like to keep it because it's a nice indication of when im panicking in a controlled fashion rather than segfaulting. but i can always add it back manually in my local set. just drop it if you dont want it, i've split it into it's own commit.

## Tests

The test runner works now! Here is the current pass rate:
```
860 passed, 1064 failed - 44.70% passing
```
(so much boilerplate to write :P)

### CODE REVIEW

**Comment by Rexicon226** [comment]
this is a good PR and all, but some of the strategies ive used aren't really clean and i dont like them. ill make this a draft while i redo some stuff.

**Comment by Rexicon226** [comment]
it really shouldn't have been that difficult to merge in a couple more commits 😄 

anyways, i've rewritten a large part of the pointer management, i'm happy enough for this to land.

File: lib/std/start.zig [hunk 1]
```diff
@@ -155,7 +155,7 @@ fn exit2(code: usize) noreturn {
                 asm volatile ("ecall"
                     :
                     : [number] "{a7}" (94),
-                      [arg1] "{a0}" (0),
```
Comment (kubkon) [inline lib/std/start.zig:158]: Yikes, good catch!

File: lib/std/start.zig [hunk 2]
```diff
@@ -61,6 +60,10 @@ comptime {
                 } else if (@typeInfo(@TypeOf(root.main)).Fn.calling_convention != .C) {
                     @export(main, .{ .name = "main" });
                 }
+            } else if (native_arch.isRISCV()) {
```
Comment (kubkon) [inline lib/std/start.zig]: I think this maybe causing the CIs to fail as this change impacts all RISC-V targets and backends.

Comment (Rexicon226) [reply lib/std/start.zig]: Yep, that was it, thanks!

**Review by kubkon — APPROVED** [review]
Great work!

**Comment by kubkon** [comment]
After you resolve conflicts, we're good to go!

**Comment by Rexicon226** [comment]
Done!

**Comment by kubkon** [comment]
Hmm, I still see conflicts.

**Comment by Rexicon226** [comment]
Yeah, my fault. Github had a bit of a moment and decided to mark the entire merge commit as belonging to this PR. I don't think we need a 132k PR :sweat: . Gimme a couple minutes I'll figure it out.

**Pull request merged**